docs/dev_guides/new_library_comparisons_and_levels.md

```diff
@@ -0,0 +1,286 @@
+
+```python
+class ArrayLengthLevelBase(ComparisonLevel):
+    def __init__(
+        self,
+        col_name: str,
+        length_difference: int,
+        m_probability=None,
+    ):
+        """Compares two arrays whose sizes are within a fixed distance
+        | length(arr_1) - length(arr_2) | <= (length_difference)
+
+        Arguments:
+            col_name (str): Input column name
+            length_difference (int): Maximum difference in array lengths
+            m_probability (float, optional): Starting value for m probability.
+                Defaults to None.
+
+        Returns:
+            ComparisonLevel: A comparison level that evaluates the size difference
+                between two arrays
+        """
+        col = InputColumn(col_name, sql_dialect=self._sql_dialect)
+        col_l, col_r = col.names_l_r()
+
+        sql_exp = (
+            f"abs("
+            f"{self._array_length_function_name}({col_l}) - "
+            f"{self._array_length_function_name}({col_r})"
+            f") <= {length_difference}"
+        )
+        level_dict = {
+            "sql_condition": sql_exp,
+            "label_for_charts": f"Array sizes differ by at most {length_difference)",
+        }
+
+        if m_probability:
+            level_dict["m_probability"] = m_probability
+
+        super().__init__(level_dict, sql_dialect=self._sql_dialect)
+```
+
+If you are using a new dialect-dependent property (as we are in this case), then it should be added as a property on `DialectBase` in [`splink.dialect_base.py`](https://github.com/moj-analytical-services/splink/blob/master/splink/dialect_base.py),
+with either a sensible default value, or raising a `NotImplementedError`:
+
+```python
+class DialectBase():
+    ...
+    @property
+    def _array_length_function_name():
+        return "array_length"
+```
+
+Then any dialects that use a different value can override this (e.g in [`splink.spark.spark_base.py`](https://github.com/moj-analytical-services/splink/blob/master/splink/spark/spark_base.py)):
+
+```python
+class SparkBase(DialectBase):
+    ...
+    @property
+    def _array_length_function_name():
+        return "array_size"
+```
+
+Then any dialects where this comparison level can be used can simply inherit from this dialect-specific base, along with the comparison level base `ArrayLengthLevelBase` - here in [`splink.spark.spark_comparison_level_library.py`](https://github.com/moj-analytical-services/splink/blob/master/splink/spark/spark_comparison_level_library.py):
+
+```python
+class array_length_level(SparkBase, ArrayLengthLevelBase):
+    pass
+```
+
+or for DuckDB in [`splink.duckdb.duckdb_comparison_level_library.py`](https://github.com/moj-analytical-services/splink/blob/master/splink/duckdb/duckdb_comparison_level_library.py):
+
+```python
+class array_length_level(DuckDBBase, ArrayLengthLevelBase):
+    pass
+```
```

Comment:
I appreciate that this is why you've added a default value in `DialectBase`.

I'd suggest adding a `NotImplementedError` and adding a default to `DuckDBBase`.

Reply:
Yeah agree, makes more sense to push people in that direction. Have amended 👍